File: examples/ospf-kotlin-example/src/main/fuookami/ospf/kotlin/example/framework_demo/demo4/flight_task_context/model/Airport.kt
package fuookami.ospf.kotlin.example.framework_demo.demo4.flight_task_context.model

import fuookami.ospf.kotlin.example.framework_demo.demo4.infrastructure.*

enum class AirportType {
    Domestic {
        override val isDomainType: Boolean get() = true
    },
    Regional,
    International;

    open val isDomainType: Boolean get() = false
}

data class Airport(
    val icao: ICAO,
    val type: AirportType,
    val base: Boolean = false
) {
    companion object {
        private val pool = HashMap<ICAO, Airport>()
        val values by pool::values

        operator fun invoke(icao: ICAO): Airport? {
            return pool[icao]
        }
    }

    init {
        pool[icao] = this
    }

    override fun hashCode(): Int {
        assert(icao.code.length == 4 && icao.code.all { it.isUpperCase() })

        var ret = 0
        for (ch in icao.code) {
            ret = ret shl 4
            ret = ret or (ch - 'A')
        }
        return ret
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Airport

        return icao == other.icao
    }

    override fun toString() = "$icao"
}

data class Route(
    val dep: Airport,
    val arr: Airport
)

enum class ArrivalAirportScope {
    Master,
    Alternate
}

File: ospf-rust-example/src/framework/demo4/domain/task/model/airport.rs
//! 机场模型模块 / Airport model module

use std::collections::HashMap;
use time::Duration;
use crate::framework::demo4::infrastructure::Icao;

/// 机场类型 / Airport type
#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, PartialOrd, Ord)]
pub enum AirportType {
    /// 国内机场 / Domestic airport
    Domestic,
    /// 区域机场 / Regional airport
    Regional,
    /// 国际机场 / International airport
    International,
}

impl AirportType {
    /// 是否为国内机场 / Check if domestic airport
    pub fn is_domain_type(&self) -> bool {
        matches!(self, AirportType::Domestic)
    }
}

/// 机场 / Airport
#[derive(Debug, Clone)]
pub struct Airport {
    /// ICAO 代码 / ICAO code
    pub icao: Icao,
    /// 机场类型 / Airport type
    pub airport_type: AirportType,
    /// 旅客过站时间 / Passenger transfer time
    pub passenger_transfer_time: Duration,
    /// 货物过站时间 / Cargo transfer time
    pub cargo_transfer_time: Duration,
    /// 是否为基地机场 / Whether this is a base airport
    pub base: bool,
}

impl PartialEq for Airport {
    fn eq(&self, other: &Self) -> bool {
        self.icao == other.icao
    }
}
impl Eq for Airport {}

impl std::hash::Hash for Airport {
    fn hash<H: std::hash::Hasher>(&self, state: &mut H) {
        self.icao.hash(state);
    }
}

/// 航线 / Route
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Route {
    /// 出发机场 / Departure airport
    pub dep: Airport,
    /// 到达机场 / Arrival airport
    pub arr: Airport,
}
